File: reconciliation-app/frontend/src/app/services/frais-transaction.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { FraisTransaction, FraisTransactionRequest } from '../models/frais-transaction.model';

@Injectable({
    providedIn: 'root'
})
export class FraisTransactionService {
    private apiUrl = `${environment.apiUrl}/frais-transaction`;

    constructor(private http: HttpClient) {}

    // Récupérer tous les frais de transaction
    getAllFraisTransactions(): Observable<FraisTransaction[]> {
        return this.http.get<FraisTransaction[]>(this.apiUrl);
    }

    // Récupérer tous les frais de transaction actifs
    getAllFraisTransactionsActifs(): Observable<FraisTransaction[]> {
        return this.http.get<FraisTransaction[]>(`${this.apiUrl}/actifs`);
    }

    // Récupérer un frais de transaction par ID
    getFraisTransactionById(id: number): Observable<FraisTransaction> {
        return this.http.get<FraisTransaction>(`${this.apiUrl}/${id}`);
    }

    // Récupérer les frais de transaction par service
    getFraisTransactionsByService(service: string): Observable<FraisTransaction[]> {
        return this.http.get<FraisTransaction[]>(`${this.apiUrl}/service/${service}`);
    }

    // Récupérer les frais de transaction par agence
    getFraisTransactionsByAgence(agence: string): Observable<FraisTransaction[]> {
        return this.http.get<FraisTransaction[]>(`${this.apiUrl}/agence/${agence}`);
    }

    // Récupérer le frais applicable pour un service et une agence
    getFraisApplicable(service: string, agence: string): Observable<FraisTransaction> {
        return this.http.get<FraisTransaction>(`${this.apiUrl}/applicable?service=${service}&agence=${agence}`);
    }

    // Récupérer tous les services uniques
    getAllServices(): Observable<string[]> {
        return this.http.get<string[]>(`${this.apiUrl}/services`);
    }

    // Récupérer toutes les agences uniques
    getAllAgences(): Observable<string[]> {
        return this.http.get<string[]>(`${this.apiUrl}/agences`);
    }

    // Exporter tous les frais de transaction
    exportFraisTransactions(): Observable<any> {
        return this.http.get<any>(`${this.apiUrl}/export`);
    }

    // Créer un nouveau frais de transaction
    createFraisTransaction(frais: FraisTransactionRequest): Observable<FraisTransaction> {
        return this.http.post<FraisTransaction>(this.apiUrl, frais);
    }

    // Mettre à jour un frais de transaction
    updateFraisTransaction(id: number, frais: FraisTransactionRequest): Observable<FraisTransaction> {
        return this.http.put<FraisTransaction>(`${this.apiUrl}/${id}`, frais);
    }

    // Supprimer un frais de transaction
    deleteFraisTransaction(id: number): Observable<boolean> {
        return this.http.delete<boolean>(`${this.apiUrl}/${id}`);
    }

    // Activer/désactiver un frais de transaction
    toggleFraisTransaction(id: number): Observable<boolean> {
        return this.http.put<boolean>(`${this.apiUrl}/${id}/toggle`, {});
    }

    // Tester le calcul des frais
    testFraisCalculation(params: any): Observable<any> {
        const queryParams = new URLSearchParams();
        Object.keys(params).forEach(key => {
            if (params[key] !== null && params[key] !== undefined) {
                queryParams.append(key, params[key].toString());
            }
        });
        return this.http.get<any>(`${this.apiUrl}/test-calculation?${queryParams.toString()}`);
    }
} 

File: reconciliation-app/backend/src/main/java/com/reconciliation/service/FraisTransactionService.java
package com.reconciliation.service;

import com.reconciliation.entity.FraisTransactionEntity;
import com.reconciliation.repository.FraisTransactionRepository;
import com.reconciliation.dto.FraisTransactionRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class FraisTransactionService {
    
    @Autowired
    private FraisTransactionRepository fraisTransactionRepository;
    
    /**
     * Créer un nouveau frais de transaction
     */
    @Transactional
    public FraisTransactionEntity createFraisTransaction(FraisTransactionRequest request) {
        // Vérifier s'il existe déjà un frais pour ce service et cette agence
        Optional<FraisTransactionEntity> existingFrais = fraisTransactionRepository.findFraisApplicable(request.getService(), request.getAgence());
        if (existingFrais.isPresent()) {
            throw new IllegalArgumentException("Un frais de transaction existe déjà pour le service '" + request.getService() + "' et l'agence '" + request.getAgence() + "'");
        }
        
        FraisTransactionEntity frais = new FraisTransactionEntity();
        frais.setService(request.getService());
        frais.setAgence(request.getAgence());
        frais.setMontantFrais(request.getMontantFrais());
        frais.setTypeCalcul(request.getTypeCalcul() != null ? request.getTypeCalcul() : "NOMINAL");
        frais.setPourcentage(request.getPourcentage());
        frais.setDescription(request.getDescription());
        frais.setActif(request.getActif() != null ? request.getActif() : true);
        frais.setDateCreation(LocalDateTime.now());
        frais.setDateModification(LocalDateTime.now());
        
        return fraisTransactionRepository.save(frais);
    }
    
    /**
     * Mettre à jour un frais de transaction existant
     */
    @Transactional
    public FraisTransactionEntity updateFraisTransaction(Long id, FraisTransactionRequest request) {
        FraisTransactionEntity frais = fraisTransactionRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Frais de transaction non trouvé avec ID: " + id));
        
        // Vérifier s'il existe déjà un autre frais pour ce service et cette agence (excluant celui en cours de modification)
        Optional<FraisTransactionEntity> existingFrais = fraisTransactionRepository.findFraisApplicable(request.getService(), request.getAgence());
        if (existingFrais.isPresent() && !existingFrais.get().getId().equals(id)) {
            throw new IllegalArgumentException("Un frais de transaction existe déjà pour le service '" + request.getService() + "' et l'agence '" + request.getAgence() + "'");
        }
        
        frais.setService(request.getService());
        frais.setAgence(request.getAgence());
        frais.setMontantFrais(request.getMontantFrais());
        frais.setTypeCalcul(request.getTypeCalcul() != null ? request.getTypeCalcul() : "NOMINAL");
        frais.setPourcentage(request.getPourcentage());
        frais.setDescription(request.getDescription());
        if (request.getActif() != null) {
            frais.setActif(request.getActif());
        }
        frais.setDateModification(LocalDateTime.now());
        
        return fraisTransactionRepository.save(frais);
    }
    
    /**
     * Récupérer un frais de transaction par ID
     */
    public Optional<FraisTransactionEntity> getFraisTransactionById(Long id) {
        return fraisTransactionRepository.findById(id);
    }
    
    /**
     * Récupérer tous les frais de transaction actifs
     */
    public List<FraisTransactionEntity> getAllFraisTransactionsActifs() {
        return fraisTransactionRepository.findByActifTrueOrderByDateModificationDesc();
    }
    
    /**
     * Récupérer tous les frais de transaction
     */
    public List<FraisTransactionEntity> getAllFraisTransactions() {
        return fraisTransactionRepository.findAllOrderByDateModificationDesc();
    }
    
    /**
     * Récupérer les frais de transaction par service
     */
    public List<FraisTransactionEntity> getFraisTransactionsByService(String service) {
        return fraisTransactionRepository.findByServiceAndActifTrueOrderByDateModificationDesc(service);
    }
    
    /**
     * Récupérer les frais de transaction par agence
     */
    public List<FraisTransactionEntity> getFraisTransactionsByAgence(String agence) {
        return fraisTransactionRepository.findByAgenceAndActifTrueOrderByDateModificationDesc(agence);
    }
    
    /**
     * Trouver le frais applicable pour un service et une agence donnés
     */
    public Optional<FraisTransactionEntity> getFraisApplicable(String service, String agence) {
        return fraisTransactionRepository.findFraisApplicable(service, agence);
    }
    
    /**
     * Récupérer tous les services uniques
     */
    public List<String> getAllServices() {
        return fraisTransactionRepository.findDistinctServices();
    }
    
    /**
     * Récupérer toutes les agences uniques
     */
    public List<String> getAllAgences() {
        return fraisTransactionRepository.findDistinctAgences();
    }
    
    /**
     * Supprimer un frais de transaction (suppression physique de la base de données)
     */
    @Transactional
    public boolean deleteFraisTransaction(Long id) {
        Optional<FraisTransactionEntity> frais = fraisTransactionRepository.findById(id);
        if (frais.isPresent()) {
            fraisTransactionRepository.deleteById(id);
            return true;
        }
        return false;
    }
    
    /**
     * Activer/désactiver un frais de transaction
     */
    @Transactional
    public boolean toggleFraisTransaction(Long id) {
        Optional<FraisTransactionEntity> frais = fraisTransactionRepository.findById(id);
        if (frais.isPresent()) {
            FraisTransactionEntity entity = frais.get();
            entity.setActif(!entity.getActif());
            entity.setDateModification(LocalDateTime.now());
            fraisTransactionRepository.save(entity);
            return true;
        }
        return false;
    }
} 